mtags/src/main/scala-3/scala/meta/internal/pc/completions/Completions.scala

```diff
@@ -64,99 +64,105 @@ class Completions(
+            /* Type might be referenced by a path over an object:
+             * ```
+             * object sample:
+             *    class X
+             * val a: samp@@le.X = ???
+             * ```
+             * ignore objects that has companion class
+             *
+             * Also ignore objects that doesn't have any members.
+             * By some reason traits might have a fake companion object(example: scala.sys.process.ProcessBuilderImpl)
+             */
+            val allowModule =
+              sym.is(Module) &&
+                (sym.companionClass == NoSymbol && sym.info.allMembers.nonEmpty)
+            allowModule
+          case Term if sym.isTerm || sym.is(Package) => true
+          case Import => true
+          case _ => false
+      end if
+    end include
+
+    def allowBracketSuffix: Boolean =
+      this match
+        case Type(hasTypeParams, _) => !hasTypeParams
+        case _ => false
+
+    def allowTemplateSuffix: Boolean =
+      this match
+        case Type(_, hasNewKw) => hasNewKw
+        case _ => false
+
+  end CursorPos
+
+  private lazy val cursorPos =
     calculateTypeInstanceAndNewPositions(path)
 
   private def calculateTypeInstanceAndNewPositions(
       path: List[Tree]
-  ): CursorPositionCondition =
+  ): CursorPos =
     path match
+      case (_: Import) :: _ => CursorPos.Import
+      case _ :: (_: Import) :: _ => CursorPos.Import
       case (head: (Select | Ident)) :: tail =>
         // https://github.com/lampepfl/dotty/issues/15750
         // due to this issue in dotty, because of which trees after typer lose information,
         // we have to calculate hasNoSquareBracket manually:
-        val hasNoSquareBracket =
+        val hasSquareBracket =
           val span: Span = head.srcPos.span
           if span.exists then
             var i = span.end
             while i < (text.length() - 1) && text(i).isWhitespace do i = i + 1
 
-            if (i < text.length()) then text(i) != '['
-            else true
+            if (i < text.length()) then text(i) == '['
+            else false
           else false
+
+        def typePos = CursorPos.Type(hasSquareBracket, hasNewKw = false)
+        def newTypePos =
+          CursorPos.Type(hasSquareBracket, hasNewKw = true)
+
         tail match
-          case (v: ValOrDefDef) :: _ =>
-            if v.tpt.sourcePos.contains(pos) then
-              CursorPositionCondition(true, false, false, hasNoSquareBracket)
-            else
-              CursorPositionCondition(false, false, false, hasNoSquareBracket)
-          case New(selectOrIdent: (Select | Ident)) :: _ =>
-            if selectOrIdent.sourcePos.contains(pos) then
-              CursorPositionCondition(false, true, false, hasNoSquareBracket)
-            else
-              CursorPositionCondition(false, false, false, hasNoSquareBracket)
-          case (a @ AppliedTypeTree(_, args)) :: _ =>
-            if args.exists(_.sourcePos.contains(pos)) then
```

Comment:
in here, it was intentional that the `if` guard was no put in the case. It creates a different logic when you add the guard to case. The `else` is different with the last `_` case that we have.

Reply:
I think the tweaking of logic in here, is the cause of not inserting square bracket in the test that is not passing.
`pprints` would be showing that.